src/runtime_src/core/edge/sw_em/generic_pcie_hal2/shim.cxx

```diff
@@ -528,8 +538,20 @@ namespace xclcpuemhal2 {
         }
         if (auto sec = xrt_core::xclbin::get_axlf_section(top, ASK_GROUP_TOPOLOGY)) {
           memTopologySize = sec->m_sectionSize;
-          memTopology = new char[memTopologySize];
-          memcpy(memTopology, xclbininmemory + sec->m_sectionOffset, memTopologySize);
+          memTopology = std::unique_ptr<char[]>(new char[memTopologySize]);
+          memcpy(memTopology.get(), xclbininmemory + sec->m_sectionOffset, memTopologySize);
+        }
+        //Extract EMULATION_DATA from XCLBIN       
+        if (auto sec = xrt_core::xclbin::get_axlf_section(top, EMULATION_DATA)) {
+          emuDataSize = sec->m_sectionSize;
+          emuData = std::unique_ptr<char[]>(new char[emuDataSize]);
+          memcpy(emuData.get(), xclbininmemory + sec->m_sectionOffset, emuDataSize);
+        }
+	      //Extract CONNECTIVITY section from XCLBIN       
+        if (auto sec = xrt_core::xclbin::get_axlf_section(top, CONNECTIVITY)) {
+          connectvitybufsize = sec->m_sectionSize;
+          connectvitybuf = std::unique_ptr<char[]>(new char[connectvitybufsize]);
+          memcpy(connectvitybuf.get(), xclbininmemory + sec->m_sectionOffset, connectvitybufsize);
```

Comment:
But by looking at all that pointer and `memcpy` spaghetti code, I wonder whether a cleaner way is not just to have something like:
```c++
std::vector<std::uint8_t> connectvitybuf { xclbininmemory + sec->m_sectionOffset, xclbininmemory + sec->m_sectionOffset + connectvitybufsize};
```

Reply:
@keryell , we can do this. but need to cast this memory to the required struct later.